core/src/main/java/org/elasticsearch/cluster/service/InternalClusterService.java

```diff
@@ -628,7 +629,61 @@ public String source() {
             logger.warn(sb.toString(), t);
             // TODO: do we want to call updateTask.onFailure here?
         }
+    }
+
+    private void safelyOnFailure(ClusterStateTaskListener listener, String source, Throwable failure) {
+        safely(
+            () -> listener.onFailure(source, failure),
+            e -> {
+                if (logger.isDebugEnabled()) {
+                    logger.debug("exception thrown by listener notifying of failure [{}] from [{}]", e, failure, source);
+                }
+            });
+    }
 
+    private void safelyOnNoLongerMaster(ClusterStateTaskListener listener, String source) {
+        safely(
+            () -> listener.onNoLongerMaster(source),
+            e -> {
+                if (logger.isDebugEnabled()) {
+                    logger.debug("exception thrown by listener while notifying no longer master from [{}]", e, source);
+                }
+            });
+    }
+
+    private void safelyClusterStateProcessed(ClusterStateTaskListener listener, String source, ClusterState oldState, ClusterState newState) {
+        safely(
+            () -> listener.clusterStateProcessed(source, oldState, newState),
+            e -> {
+                if (logger.isDebugEnabled()) {
+                    logger.debug(
+                        "exception thrown by listener while notifying of cluster state processed from [{}], old cluster state:\n{}\nnew cluster state:\n{}",
+                        e,
+                        source,
+                        oldState.prettyPrint(),
+                        newState.prettyPrint());
+                }
+            });
+    }
+
+    private <T> void safelyClusterStatePublished(ClusterStateTaskExecutor<T> executor, ClusterState newState) {
+        safely(
+            () -> executor.clusterStatePublished(newState),
+            e -> {
+                if (logger.isDebugEnabled()) {
+                    logger.debug("exception thrown by executor while notifying of cluster state published from [{}]", e, newState.prettyPrint());
+                }
+            });
+    }
+
+    private void safely(Runnable runnable, Consumer<Exception> consumer) {
+        try {
+            runnable.run();
+        } catch (Exception e) {
+            try {
+                consumer.accept(e);
+            } catch (Exception ignored) {}
```

Comment:
It feels squicky to ignore the exception. Personally I'd `assert false : "Exceptions not expected here."` and `logger.error` about it.